win: implement is_pmem_proc

ChandKV opened the pull request:
1. Implement SortedQ using CircularQ
2. Add tests for SortedQ in win_lists
3. Convert FileMappingList to a SortedQ
   a. Fix the PVOID to PVOID \* conversion
4. Implement setenv and unsetenv
5. Add tests for setenv and unsetenv
6. Implement is_pmem_proc
7. Port tests for pmem_is_pmem to test is_pmem_proc
   a. Define PMEM_IS_PMEM_FORCE_TEST environment variable to test
      is_pmem_proc with a non-pmem volume
8. Add Windows specific test for pmem_is_pmem to check the
   interleaved file mapping case using mmap and MapViewOfFile
9. Fix obj_pool_lookup to generate correct DebugInformation
   1. Fix bugs in mmap_split's error handling path

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1115)

<!-- Reviewable:end -->

krzycz commented:
A general comment - we recommend to squash commits in a pull request, but only if they belong to the same logical change. In particular, we try to avoid commits like "win: fix typo/cstyle" in a pull request.
In this case however, I'd prefer to split this into a few separate commits (but still one pull request) since SortedQ, setenv and pmem_is_pmem are in practice separate, self-contained changes.

---

Reviewed 45 of 48 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/libpmem/pmem.c, line 494 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL3nADpdo-RwIi5soc:-KQL3nADpdo-RwIi5sod:b-q5q2a3) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/libpmem/pmem.c#L494)):*
> ```C
> 			LOG(4, "PMEM_IS_PMEM_FORCE=%d", val);
> 
> 		} else if ((ptr = getenv("PMEM_IS_PMEM_FORCE_TEST"))) {
> ```

What's the difference between PMEM_IS_PMEM_FORCE and PMEM_IS_PMEM_FORCE_TEST?
Do we really need yet another env variable?

---

*[src/libpmem/pmem_windows.c, line 98 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ7csE44wWXXAe-0eW:-KQQ7csE44wWXXAe-0eX:b-c905te) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/libpmem/pmem_windows.c#L98)):*
> ```C
> 	SORTEDQ_FOREACH(mt, &FileMappingQHead, ListEntry) {
> 		if (mt->BaseAddress >= end) {
> 			LOG(3, "ignoring all mapped ranges beyond given range");
> ```

I would use level 4 here and below.

---

*[src/test/pmem_is_pmem/TEST0.PS1, line 59 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL5ejtguGtooGivrcS:-KQL5ejtguGtooGivrcT:bghxr81) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem/TEST0.PS1#L59)):*
> ```
> 
> create_holey_file 2M $DIR\testfile1
> $ENV:PMEM_IS_PMEM_FORCE_TEST = "0"
> ```

Why do we need this?

---

*[src/test/pmem_is_pmem_windows/pmem_is_pmem_windows.c, line 35 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ8Vqlc4Rtuf2bJMhx:-KQQ8Vqlc4Rtuf2bJMhy:b-e64cua) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem_windows/pmem_is_pmem_windows.c#L35)):*
> ```C
> 
> /*
>  * pmem_is_pmem.c -- unit test for pmem_is_pmem()
> ```

pmem_is_pmem_windows.c

---

*[src/windows/win_mmap.c, line 425 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ4zq_15t4z_XwqmYE:-KQQ4zq_15t4z_XwqmYF:b-ze1f1h) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/windows/win_mmap.c#L425)):*
> ```C
> 			LOG(4, "file is not DAX mapped - handle: %p", fh);
> 	} else
> 		LOG(4, "failed to query volume information : %08x",
> ```

What about ERR() ?

---

*[src/windows/win_mmap.c, line 448 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ57Ru_3Us9PKpullx:-KQQ57Ru_3Us9PKpully:b-ucqzow) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/windows/win_mmap.c#L448)):*
> ```C
> 	/*
> 	 * In this routine we copy flags from mt to the two subsets that we
> 	 * create.  All flags may not be safe to propagat so let's assert
> ```

propagate, appropriate ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115#-:-KQL16EVP0aDQq5738I0:bn5tr2u)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Sure, I too would prefer multiple commits but was thinking that I'm following NVML guidelines. Once all the comments are resolved, let me split this pull request into multiple commits and send a pull request.

---

Review status: 44 of 48 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/libpmem/pmem.c, line 494 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL3nADpdo-RwIi5soc:-KQT8PXasUZjlPaNz5Gu:b-oxnjsq) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/libpmem/pmem.c#L494)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> What's the difference between PMEM_IS_PMEM_FORCE and PMEM_IS_PMEM_FORCE_TEST?
> Do we really need yet another env variable?
</details>
I'm open different name suggestion. PMEM_IS_PMEM_FORCE_TEST would let pmem_is_pmem use is_pmem_proc() instead of is_pmem_always(). It forces the decision that the page or mapped region is DAX backed at the lowest level after the libpmem calls into OS API to decide if the page / region is DAX mapped. This way we could test this code in non-pmem volume.

---

*[src/libpmem/pmem_windows.c, line 98 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ7csE44wWXXAe-0eW:-KQT8Mw82Ty1vVIo7WBO:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/libpmem/pmem_windows.c#L98)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I would use level 4 here and below.
</details>
Done.

---

*[src/test/pmem_is_pmem/TEST0.PS1, line 59 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL5ejtguGtooGivrcS:-KQT7YmKRrtXYyCoaZA4:b58kmte) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem/TEST0.PS1#L59)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Why do we need this?
</details>
To execute the code that iterates through the tracked regions in the list and iterates through each page in the untracked regions and decide if all are DAX mapped. This flags affects the decision there. So we could test the code path in non-pmem volume.

---

*[src/test/pmem_is_pmem_windows/pmem_is_pmem_windows.c, line 35 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ8Vqlc4Rtuf2bJMhx:-KQT7UfAHvsQt6GN178L:b-i2f9eo) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem_windows/pmem_is_pmem_windows.c#L35)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> pmem_is_pmem_windows.c
</details>
:) Done.

---

*[src/windows/win_mmap.c, line 425 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ4zq_15t4z_XwqmYE:-KQT75IMJnxEKErYz5aC:bgnoagf) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/windows/win_mmap.c#L425)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> What about ERR() ?
</details>
Done. Picked ERR().

---

*[src/windows/win_mmap.c, line 448 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQQ57Ru_3Us9PKpullx:-KQT6eAJK7C9_34ZZsog:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/windows/win_mmap.c#L448)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> propagate, appropriate ?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115#-:-KQT96sm2U2e2WAAzWVc:b-d8shyr)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/test/pmem_is_pmem/TEST0.PS1, line 59 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL5ejtguGtooGivrcS:-KQX22tBikZ60rPZqouW:b-vwijh9) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem/TEST0.PS1#L59)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> To execute the code that iterates through the tracked regions in the list and iterates through each page in the untracked regions and decide if all are DAX mapped. This flags affects the decision there. So we could test the code path in non-pmem volume.
</details>
Oh, I think I got what you mean. But I think that introducing a new env var just to satisfy one test is not the right approach.
The unit test for Windows version of is_pmem_proc() should rather:
1) link pmem_windows.c (perhaps also win_mmap.c) to the test program (no need to link libpmem.dll);
2) prepare a test set of memory regions (pmme/non-pmem, contiguous/non-contiguous, ...) in a mapping list/queue (FileMappingQHead) - there's no need to do any real mappings - just insert test data directly to the mapping list;
3) call is_pmem_proc() directly - not via pmem_is_pmem().

In the next step, we could also hook/mock the call to MM API and simulate it returns true/false when querying on whether the given page is DAX-mapped or not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/test/pmem_is_pmem/TEST0.PS1, line 59 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL5ejtguGtooGivrcS:-KQXLAW33RPoR91MpaCv:b-bx7gp1) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem/TEST0.PS1#L59)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Oh, I think I got what you mean. But I think that introducing a new env var just to satisfy one test is not the right approach.
> The unit test for Windows version of is_pmem_proc() should rather:
> 1) link pmem_windows.c (perhaps also win_mmap.c) to the test program (no need to link libpmem.dll);
> 2) prepare a test set of memory regions (pmme/non-pmem, contiguous/non-contiguous, ...) in a mapping list/queue (FileMappingQHead) - there's no need to do any real mappings - just insert test data directly to the mapping list;
> 3) call is_pmem_proc() directly - not via pmem_is_pmem().
> 
> In the next step, we could also hook/mock the call to MM API and simulate it returns true/false when querying on whether the given page is DAX-mapped or not.
> 
</details>
I see what you say. Yes PMEM_IS_PMEM_FORCE_TEST is essentially acting like hooking MM and FS calls. I was modelling it based on PMEM_IS_PMEM_FORCE. Why do we have that?

I'd prefer hooking MM and FS API, let me attempt the same. I'm planning to follow @JLLLinn model of having function pointers and then exporting a function that would set those function pointers, so that test can set it. This set_*() function would be exported only in DEBUG configuration. How did you like this, approach?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/test/pmem_is_pmem/TEST0.PS1, line 59 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL5ejtguGtooGivrcS:-KQavsYrx71KxdZPEr8L:b9pvwit) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem/TEST0.PS1#L59)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> I see what you say. Yes PMEM_IS_PMEM_FORCE_TEST is essentially acting like hooking MM and FS calls. I was modelling it based on PMEM_IS_PMEM_FORCE. Why do we have that?
> 
> I'd prefer hooking MM and FS API, let me attempt the same. I'm planning to follow @JLLLinn model of having function pointers and then exporting a function that would set those function pointers, so that test can set it. This set_*() function would be exported only in DEBUG configuration. How did you like this, approach?
</details>
Based on our chat I believe the plan is to use 2 for PMEM_IS_PMEM_FORCE to hint this. Let me know you have concerns with that and we can land on something better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 40 of 48 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/test/pmem_is_pmem/TEST0.PS1, line 59 \[r2\]](https://reviewable.io:443/reviews/pmem/nvml/1115#-KQL5ejtguGtooGivrcS:-KR4UfyEBCk1EMeHbV3H:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/225a337e52f082138abc478bcd885426045b4492/src/test/pmem_is_pmem/TEST0.PS1#L59)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> Based on our chat I believe the plan is to use 2 for PMEM_IS_PMEM_FORCE to hint this. Let me know you have concerns with that and we can land on something better.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/libpmem/pmem.c, line 495 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRG0qFXgx1EcKHo_2im:-KRG0qFXgx1EcKHo_2in:b-hlajqx) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem.c#L495)):*
> ```C
> 				break;
> 			case 2:
> 				Pmem_is_pmem_force_test = (val != 0);
> ```

If val == 2, then it's not zero, for sure... ;-)

---

*[src/libpmem/pmem_windows.c, line 65 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZF_v3UJ_2hDNAYxAC:-KRZF_v5DyYWlL6pPJNK:bk0e2dn) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L65)):*
> ```C
> 		 * is not direct mapped.
> 		 */
> 		retval = 0;
> ```

Wouldn't it be better to put this check here (at the lowest level), not in *is_pmem_proc()*?:
```c
if (Pmem_is_pmem_force_test)
   retval = 1;
else
  ...
```

---

*[src/libpmem/pmem_windows.c, line 109 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZEgT0w45Po0U9g3AJ:-KRZEgT1iOTLfGj7nDXb:b-8cal4n) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L109)):*
> ```C
> 				mt->BaseAddress, mt->EndAddress);
> 			retval = 0;
> 			if (Pmem_is_pmem_force_test) {
> ```

So, if this is to override "Flags", then why not to make this change somewhere in **win_mmap.c**, right after the call to *GetVolumeInformationByHandleW*?
I thought the whole idea was to simulate that *GetVolumeInformationByHandleW* returns FILE_MAPPING_TRACKER_FLAG_DIRECT_MAPPED flag, and/or that MM API (not available yet) says that given page is direct-mapped.
So, IMHO this 'if' should be close to those function calls.

Also, I'm still not convinced that we need a new flag/new value of env var, just to test this function.
On Linux, to test is_pmem_proc() function (which is actually a "/proc/self/smaps" file parser), we feed it with handcrafted "smaps" files (we need to hook fopen() for that).  Here, we could prepare a handcrafted FileMapping list (with some ranges "direct-mapped" and some w/o this flag), and then call is_pmem_proc(). 

So, *pmem_is_pmem_windows* should be more like Windows version of *pmem_is_pmem_proc* test. Perhaps we could even rename the existing tests to have:
- pmem_is_pmem (same for Linux and Windows - only testing PMEM_IS_PMEM_FORCE = 0 or 1)
- pmem_is_pmem_proc_linux (currently pmem_is_pmem_proc)
- pmem_is_pmem_proc_windows (currently pmem_is_pmem_windows)

For the latter, I'd link statically to THIS file (not the entire libpmem) and in the test program I'd prepare some dummy FileMapping list, then call is_pmem_proc() directly.
Does it make sense?

---

*[src/test/pmem_is_pmem/TEST1.PS1, line 60 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRG1V7er9KGj9DSVFDK:-KRG1V7iR6jb8ygud-t_:b9mk4pm) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/test/pmem_is_pmem/TEST1.PS1#L60)):*
> ```
> create_holey_file 2M $DIR\testfile1
> 
> $SAVED_PMEM_IS_PMEM_FORCE = $ENV:PMEM_IS_PMEM_FORCE
> ```

Hmmm, I didn't think it might be necessary to save/restore the old value of the env variable(s). 
On Linux, the change only affects the current process (test), so there is no need to restore old value. Is it different on Windows, or is it because of how TEST*.PS1 scripts are executed (not as a sub-process)?
I mean, if this is required for THIS test, then probably also for other tests!!! Would be good to verify.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/libpmem/pmem.c, line 495 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRG0qFXgx1EcKHo_2im:-KRZXJFsYnIDmhT1hhFa:b8lf3s2) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem.c#L495)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> If val == 2, then it's not zero, for sure... ;-)
</details>
Sure :). Fixed. Let me update the PR after making the below changes as well.

---

*[src/libpmem/pmem_windows.c, line 65 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZF_v3UJ_2hDNAYxAC:-KRZXjNZ0Rh9OJafVOiP:bqqf7jo) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Wouldn't it be better to put this check here (at the lowest level), not in *is_pmem_proc()*?:
> ```c
> if (Pmem_is_pmem_force_test)
>    retval = 1;
> else
>   ...
> ```
> 
</details>
I think it's better to let one single function act on a test-hook instead of spreading it across if reasonably possible. I was preferring localizing the hook checks to reducing the number of checks. This is subjective, if you prefer the other way I can change it.

---

*[src/libpmem/pmem_windows.c, line 109 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZEgT0w45Po0U9g3AJ:-KRZal6VMRn9Ycuspdja:b-8h1kk7) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L109)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> So, if this is to override "Flags", then why not to make this change somewhere in **win_mmap.c**, right after the call to *GetVolumeInformationByHandleW*?
> I thought the whole idea was to simulate that *GetVolumeInformationByHandleW* returns FILE_MAPPING_TRACKER_FLAG_DIRECT_MAPPED flag, and/or that MM API (not available yet) says that given page is direct-mapped.
> So, IMHO this 'if' should be close to those function calls.
> 
> Also, I'm still not convinced that we need a new flag/new value of env var, just to test this function.
> On Linux, to test is_pmem_proc() function (which is actually a "/proc/self/smaps" file parser), we feed it with handcrafted "smaps" files (we need to hook fopen() for that).  Here, we could prepare a handcrafted FileMapping list (with some ranges "direct-mapped" and some w/o this flag), and then call is_pmem_proc(). 
> 
> So, *pmem_is_pmem_windows* should be more like Windows version of *pmem_is_pmem_proc* test. Perhaps we could even rename the existing tests to have:
> - pmem_is_pmem (same for Linux and Windows - only testing PMEM_IS_PMEM_FORCE = 0 or 1)
> - pmem_is_pmem_proc_linux (currently pmem_is_pmem_proc)
> - pmem_is_pmem_proc_windows (currently pmem_is_pmem_windows)
> 
> For the latter, I'd link statically to THIS file (not the entire libpmem) and in the test program I'd prepare some dummy FileMapping list, then call is_pmem_proc() directly.
> Does it make sense?
> 
> 
</details>
I didn't change the flag, as the intention is to just let this function behave as if the flag is not set. Again to have a single function act on the test-hook.

You might like this, I think that I could appreciate your approach as well. Let me try that.

By the way it's interesting to see the name suggestion, I was thinking that pmem_is_pmem_proc is named so because you are hooking the open to a file in /proc similar to util_map_proc. Now I could see how this is named. Let me change them accordingly.

---

*[src/test/pmem_is_pmem/TEST1.PS1, line 60 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRG1V7er9KGj9DSVFDK:-KRZ_Ixjk_gujjw8b9eq:b-nnaslb) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/test/pmem_is_pmem/TEST1.PS1#L60)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Hmmm, I didn't think it might be necessary to save/restore the old value of the env variable(s). 
> On Linux, the change only affects the current process (test), so there is no need to restore old value. Is it different on Windows, or is it because of how TEST*.PS1 scripts are executed (not as a sub-process)?
> I mean, if this is required for THIS test, then probably also for other tests!!! Would be good to verify.
</details>
If we go the RUNTESTS route this is not required, as we launch the TEST*.PS1 as a separate process. But if we run TEST*.PS1 directly, then this might be safer. We may choose to not do this and leave to the developer to set the variables correctly before running a TEST individually, thought this is better.

However, with Paul's change to run an individual test in Windows using RUNTEST, one might not need to run TEST*.PS1 directly, but piece of me still thinks that it might be easier to debug the TESTs if we run the TEST*.PS1. But not sure if we want to make sure that all tests are clean, may be we could clean this directory. Thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/libpmem/pmem_windows.c, line 65 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZF_v3UJ_2hDNAYxAC:-KSq8CXKLRrJGju58p4s:b-dklovy) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L65)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> I think it's better to let one single function act on a test-hook instead of spreading it across if reasonably possible. I was preferring localizing the hook checks to reducing the number of checks. This is subjective, if you prefer the other way I can change it.
</details>
Well, to be honest, I'd prefer the other way...

---

*[src/libpmem/pmem_windows.c, line 109 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZEgT0w45Po0U9g3AJ:-KSq8ZeqJWLFcKEJVm39:b-hk4ew1) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L109)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> I didn't change the flag, as the intention is to just let this function behave as if the flag is not set. Again to have a single function act on the test-hook.
> 
> You might like this, I think that I could appreciate your approach as well. Let me try that.
> 
> By the way it's interesting to see the name suggestion, I was thinking that pmem_is_pmem_proc is named so because you are hooking the open to a file in /proc similar to util_map_proc. Now I could see how this is named. Let me change them accordingly.
</details>
I don't think I understand that. If the flag is not set, the condition is true, so we'll enter this if() and we set *retval* to 0. But because PMEM_IS_PMEM_FORCE is set, the *retval* is immediately overwritten with "1". After all, the behavior is exactly the same as if the  FILE_MAPPING_TRACKER_FLAG_DIRECT_MAPPED flag is set in mmap() function (except the trace logs). Isn't it?

---

*[src/test/pmem_is_pmem/TEST1.PS1, line 60 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRG1V7er9KGj9DSVFDK:-KSHUFqnoWSm64knWYdT:b-5lx4mz) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/test/pmem_is_pmem/TEST1.PS1#L60)):*
<details><summary><i>Previously, ChandKV (Chandra Kumar) wrote…</i></summary>
> If we go the RUNTESTS route this is not required, as we launch the TEST*.PS1 as a separate process. But if we run TEST*.PS1 directly, then this might be safer. We may choose to not do this and leave to the developer to set the variables correctly before running a TEST individually, thought this is better.
> 
> However, with Paul's change to run an individual test in Windows using RUNTEST, one might not need to run TEST*.PS1 directly, but piece of me still thinks that it might be easier to debug the TESTs if we run the TEST*.PS1. But not sure if we want to make sure that all tests are clean, may be we could clean this directory. Thoughts?
</details>
Hmmm, I think you're right. Perhaps this is not critical, but I think we could add some save_env_vars()/restore_env_vars() fiunctions that would be executed at the beginning, and at the exit of each TEST*.PS1 script (could be a part of setup() and check()/pass() respectively). The only problem is that if the test fails, and the script terminates before it reaches check/pass, the old variables won't be restored. (BTW, I think the same problem is in your code - if expect_normal_exit fails, PMEM_IS_PMEM_FORCE would remain 2.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 240 of 253 files at r5.
Review status: 245 of 252 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 9 of 253 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c, line 42 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjOLfYIZ8_0VowNKb3:-KVjOLfYIZ8_0VowNKb4:b-ibskup) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c#L42)):*
> ```C
> #include "unittest.h"
> #include "pmem.h"
> #include "queue.h"
> ```

`#include <sys/queue.h>`

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c, line 60 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjsZL2we7hdHaACsZo:-KVjsZL2we7hdHaACsZp:b-eerkjb) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c#L60)):*
> ```C
> 
> enum test_mmap_scenarios
> get_mmap_scenarios(char *name)
> ```

Well, it's not bad, but in practice it would be enough to use one-letter args ('n' for "nothing", 'b' for "begin", etc.), which makes it much easier to parse (no strcmp). I.e.:
```
switch (argv[1][0]) {
case 'n':  return TEST_MMAP_SCENARIO_NO_HOLE;
...
```

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c, line 98 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjw34WGFnY841WCp0d:-KVjw34XYL-8SaI9yoke:b-n729wd) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c#L98)):*
> ```C
> 		*ret = 1;
> 	else
> 		*ret =  is_pmem_proc(Addr, Size);
> ```

Why not just call pmem_is_pmem()? It would handle PMEM_IS_PMEM_FORCE.
Hmmm, you'd probably link entire libpmem statically.

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.vcxproj, line 46 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjNuho_r51-s8-KJoR:-KVjNuhp7uTxioOfbDpd:b-oiyg4h) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.vcxproj#L46)):*
> ```XML
>   <PropertyGroup Label="UserMacros" />
>   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
>     <IncludePath>$(SolutionDir)\common;$(SolutionDir)\test\unittest;$(SolutionDir)\windows\include;$(SolutionDir)\include;$(SolutionDir)\libpmem;$(SolutionDir)\windows\include\sys;$(IncludePath)</IncludePath>
> ```

I don't think there is a need for additional include dir. I'd rather do `#include <sys/queue.h>`, which is how we do it on Linux.

---

*[src/windows/include/platform.h, line 75 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjHY_gdlOubxMAWm4D:-KVjHY_gdlOubxMAWm4E:bqvfbsw) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/windows/include/platform.h#L75)):*
> ```objc
> 
> /* XXX: define the definitions that will be available soon */
> #define FILE_DAX_VOLUME		0x20000000
> ```

IMHO, this is not required. It's already defined in SDK 10.0.14393.0.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

ChandKV commented:
Review status: 238 of 252 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/libpmem/pmem_windows.c, line 65 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZF_v3UJ_2hDNAYxAC:-KVk3o3lCqDkDFsAAtod:b-x16von) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Well, to be honest, I'd prefer the other way...
</details>
This variable is no longer used, as I switched you to your suggestion of unit testing by importing code. So let's chat next time :).

---

*[src/libpmem/pmem_windows.c, line 109 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRZEgT0w45Po0U9g3AJ:-KVk4_EIvMl32fMllWD0:b-f2d9hp) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/libpmem/pmem_windows.c#L109)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I don't think I understand that. If the flag is not set, the condition is true, so we'll enter this if() and we set *retval* to 0. But because PMEM_IS_PMEM_FORCE is set, the *retval* is immediately overwritten with "1". After all, the behavior is exactly the same as if the  FILE_MAPPING_TRACKER_FLAG_DIRECT_MAPPED flag is set in mmap() function (except the trace logs). Isn't it?
</details>
I see the disconnect, the idea is not just for changing what GetVolumeInformationByHandleW () returns but also to impact what the MM API would return that's why I have include the check in is_direct_mapped as well. Also I wanted to make sure that we iterate through all the pages even if this is set, so that the code behave as if the flag is not set and we are calling MM API.

Anyway, all this is gone with the new unit test approach. But I'm open to chat if you like to.

---

*[src/test/pmem_is_pmem/TEST1.PS1, line 60 at r4](https://reviewable.io:443/reviews/pmem/nvml/1115#-KRG1V7er9KGj9DSVFDK:-KVk5BVcNnwNcjg6knFr:b8a9dp5) ([raw file](https://github.com/pmem/nvml/blob/780a7a258c241243829c7f3d54c7a2f91ae2100c/src/test/pmem_is_pmem/TEST1.PS1#L60)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Hmmm, I think you're right. Perhaps this is not critical, but I think we could add some save_env_vars()/restore_env_vars() fiunctions that would be executed at the beginning, and at the exit of each TEST*.PS1 script (could be a part of setup() and check()/pass() respectively). The only problem is that if the test fails, and the script terminates before it reaches check/pass, the old variables won't be restored. (BTW, I think the same problem is in your code - if expect_normal_exit fails, PMEM_IS_PMEM_FORCE would remain 2.)
</details>
Yes we can make it like a frame work, let's defer it for now, unless you think otherwise.

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c, line 42 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjOLfYIZ8_0VowNKb3:-KVk5d6jtrvQwTLblzAL:bz25wxn) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c#L42)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> `#include <sys/queue.h>`
</details>
Done. Also I remove the the "sys" folder from the "Include Directories" and added a note in trello about the same in "Steps to configure your VS project"

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c, line 60 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjsZL2we7hdHaACsZo:-KVk8eN_yWUbpdcWhxme:bdy314r) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c#L60)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Well, it's not bad, but in practice it would be enough to use one-letter args ('n' for "nothing", 'b' for "begin", etc.), which makes it much easier to parse (no strcmp). I.e.:
> ```
> switch (argv[1][0]) {
> case 'n':  return TEST_MMAP_SCENARIO_NO_HOLE;
> ...
> ```
</details>
Unless for a performance critical path, I would prefer full names. Alternatively I can print the complete names in help and check for just the first letters as the remaining is anyway different but not a big fan of that. I assume the current implementation has your blessings, let me know if you like me to change.

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c, line 98 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjw34WGFnY841WCp0d:-KVkEuIrjCgBIsPsTDl_:b8l4q5e) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.c#L98)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Why not just call pmem_is_pmem()? It would handle PMEM_IS_PMEM_FORCE.
> Hmmm, you'd probably link entire libpmem statically.
</details>
Because I want to call the is_pmem_proc that's in the test binary and not the one in the libpmem library. It is important because the test defines FileMappingQHead and crafts the same. Alternatively I could pull pmem.c into this project but then it's not so isolated and requires other files. And I found checking the ENV locally to be more elegant.

Also as a bonus, what the unit test tests is more explicit.

I included comments here to discuss the same.

---

*[src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.vcxproj, line 46 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjNuho_r51-s8-KJoR:-KVkI8bw2zJuk_vrUdOe:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/test/pmem_is_pmem_proc_windows/pmem_is_pmem_proc_windows.vcxproj#L46)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I don't think there is a need for additional include dir. I'd rather do `#include <sys/queue.h>`, which is how we do it on Linux.
</details>
Done.

---

*[src/windows/include/platform.h, line 75 at r5](https://reviewable.io:443/reviews/pmem/nvml/1115#-KVjHY_gdlOubxMAWm4D:-KVkIBZG0yuHiiNHKo7H:b-bh7p92) ([raw file](https://github.com/pmem/nvml/blob/a8cb6b27decbd24eb9551ae6f534964092828c02/src/windows/include/platform.h#L75)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> IMHO, this is not required. It's already defined in SDK 10.0.14393.0.
</details>
:) thanks for catching that. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 253 files at r5, 14 of 14 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1115#-:-KVy_-ymNt0I2Ae4I-6M:bnfp4nl)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Merged by hand.
